### ISSUE TITLE
NPE in io.vertx.core.http.impl.HttpClientImpl

### ISSUE THREAD

**Issue by bolbat** [eclipse-vertx/vert.x#4220]
Hello

We get rare NPE's on high load in Http1xClientConnection
Vert.x: 4.2.3
Stacktrace:
```
java.lang.NullPointerException: null
	at io.vertx.core.http.impl.HttpClientImpl.lambda$doRequest$8(HttpClientImpl.java:631)
	at io.vertx.core.net.impl.pool.Endpoint.lambda$getConnection$0(Endpoint.java:52)
	at io.vertx.core.http.impl.ClientHttpEndpointBase.lambda$requestConnection$0(ClientHttpEndpointBase.java:46)
	at io.vertx.core.http.impl.SharedClientHttpStreamEndpoint$Request.handle(SharedClientHttpStreamEndpoint.java:162)
	at io.vertx.core.http.impl.SharedClientHttpStreamEndpoint$Request.handle(SharedClientHttpStreamEndpoint.java:123)
	at io.vertx.core.impl.EventLoopContext.emit(EventLoopContext.java:50)
	at io.vertx.core.impl.EventLoopContext.lambda$emit$1(EventLoopContext.java:57)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:469)
	at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:384)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:986)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:829)
```

**Comment by vietj** [comment]
do you have a reproducer ?

**Comment by bolbat** [comment]
No, this happens rarely on a high loaded instance (30-50k RPS)
I've found this in logs after migration from latest Vert.x 3.x to 4.x

**Comment by vietj** [comment]
I'm assuming you are sharing an HTTP client between verticles ?

it seems that the connection is closed when the client gets the connection from the pool.

I will investigate thanks for reporting this.

**Comment by bolbat** [comment]
Yes, WebClient shared between verticles.

**Comment by vietj** [comment]
the client does get a connection that is closed from the pool, so the request should be failed as a normal failure (connection closed) and instead there is a bug that shows this NPE.

avoiding to share would avoid this as temporary workaround for this bug.

we will fix this bug, so the client gets a proper failure instead of having this internal NPE.

**Comment by bolbat** [comment]
Also found this one NPE (other case)
```
java.lang.NullPointerException: null
	at io.vertx.core.net.impl.pool.SimpleConnectionPool$Recycle$1.run(SimpleConnectionPool.java:714)
	at io.vertx.core.net.impl.pool.CombinerExecutor.submit(CombinerExecutor.java:50)
	at io.vertx.core.net.impl.pool.SimpleConnectionPool.execute(SimpleConnectionPool.java:240)
	at io.vertx.core.net.impl.pool.SimpleConnectionPool.recycle(SimpleConnectionPool.java:730)
	at io.vertx.core.net.impl.pool.SimpleConnectionPool.access$3300(SimpleConnectionPool.java:80)
	at io.vertx.core.net.impl.pool.SimpleConnectionPool$LeaseImpl.recycle(SimpleConnectionPool.java:690)
	at io.vertx.core.http.impl.HttpClientImpl.lambda$null$6(HttpClientImpl.java:635)
	at io.vertx.core.http.impl.Http1xClientConnection$StreamImpl.tryClose(Http1xClientConnection.java:633)
	at io.vertx.core.http.impl.Http1xClientConnection$StreamImpl.handleClosed(Http1xClientConnection.java:623)
	at io.vertx.core.impl.EventLoopContext.execute(EventLoopContext.java:81)
	at io.vertx.core.impl.ContextImpl.execute(ContextImpl.java:260)
	at io.vertx.core.impl.EventLoopContext.execute(EventLoopContext.java:22)
	at io.vertx.core.http.impl.Http1xClientConnection$StreamImpl._reset(Http1xClientConnection.java:554)
	at io.vertx.core.http.impl.Http1xClientConnection$StreamImpl.reset(Http1xClientConnection.java:544)
	at io.vertx.core.http.impl.HttpClientRequestImpl.reset(HttpClientRequestImpl.java:256)
	at io.vertx.core.http.impl.HttpClientRequestBase.handleTimeout(HttpClientRequestBase.java:213)
	at io.vertx.core.http.impl.HttpClientRequestBase.lambda$setTimeout$1(HttpClientRequestBase.java:155)
	at io.vertx.core.impl.VertxImpl$InternalTimerHandler.handle(VertxImpl.java:893)
	at io.vertx.core.impl.VertxImpl$InternalTimerHandler.handle(VertxImpl.java:860)
	at io.vertx.core.impl.EventLoopContext.emit(EventLoopContext.java:50)
	at io.vertx.core.impl.ContextImpl.emit(ContextImpl.java:274)
	at io.vertx.core.impl.EventLoopContext.emit(EventLoopContext.java:22)
	at io.vertx.core.impl.AbstractContext.emit(AbstractContext.java:53)
	at io.vertx.core.impl.EventLoopContext.emit(EventLoopContext.java:22)
	at io.vertx.core.impl.VertxImpl$InternalTimerHandler.run(VertxImpl.java:883)
	at io.netty.util.concurrent.PromiseTask.runTask(PromiseTask.java:98)
	at io.netty.util.concurrent.ScheduledFutureTask.run(ScheduledFutureTask.java:170)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:469)
	at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:384)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:986)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:829)
```

**Comment by vietj** [comment]
this other NPE might be indeed another case

**Comment by bolbat** [comment]
I can create another issue if needed

**Comment by vietj** [comment]
no it's fine

**Comment by vietj** [comment]
I will investigate this new case.

**Comment by vietj** [comment]
I think I found the issue for the second case

**Comment by vietj** [comment]
is it possible you test snapshots of vertx core with the fixes ?

**Comment by bolbat** [comment]
Problem that i see this exceptions only at production on high load
We are using automated CI pipeline with maven build and managed dependencies
I can test fixed version if it will be available from some "snapshots" repository
But also want to mention that this exceptions are rare, happening even not each hour

**Comment by vietj** [comment]
those exception are rare and hard to reproduce, I look at the issue and find an issue I believe is corresponding and then a new snapshot will be available. This is to ensure that what we fix works in practice in your env.

we can handle this privately and I can provide a better snapshots (i.e your current version with issue cherry picked)  to ensure there is not drift in your prod env.

**Comment by vietj** [comment]
see https://github.com/eclipse-vertx/vert.x/issues/4224

**Comment by bolbat** [comment]
Seems that this issue (both NPE's) was resolved in 4.2.4-SNAPSHOT

**Comment by vietj** [comment]
thanks for reporting
